### PR TITLE
Remove default detectors IDs from Workspace2D

### PR DESCRIPTION
**Pull request by SimonHeybrock**
`Workspace2D` used to init its spectra with default detector IDs starting from 1. This does not make sense since there is not instrument, i.e., the IDs are just dangling references, and if an instrument is loaded IDs are in general not contiguous and starting from 1, yielding orphaned detectors and still dangling references.

Most actually code rebuilds the spectra mapping upon loading an instrument so this is mainly relevant for unit tests. Note that `EventWorkspace` does not set default IDs, so this PR brings `Workspace2D` in line with `EventWorkspace`.

As a side effect:
- Fixed bug introduced when refactoring `LoadQKK` (would not init histograms anymore).
- Fixed `UpdateInstrumentFromFileTest`, which did not set IDs for any but spectrum 0.

**To test:**

Code review.
No associated issue.
No release notes, this should be an internal change?

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/AsymmetryCalc.cpp [hunk 1]
```diff
@@ -129,6 +129,7 @@ void AsymmetryCalc::exec() {
   // Create a point data workspace with only one spectra for forward
   auto outputWS = DataObjects::create<API::HistoWorkspace>(
       *inputWS, 1, tmpWS->points(forward));
+  outputWS->getSpectrum(0).setDetectorID(static_cast<detid_t>(1));
```
Comment (OwenArnold) [inline Framework/Algorithms/src/AsymmetryCalc.cpp]: Isn't the modern way to do this via `setSpectrumDefinition` on `ExperimentInfo`?

File: Framework/Algorithms/src/CreateWorkspace.cpp [hunk 2]
```diff
@@ -199,13 +199,6 @@ void CreateWorkspace::exec() {
     const std::vector<double>::difference_type yEnd = yStart + ySize;
     auto local_i = localIndices[0];
 
-    if (!parentWS) {
-      // There is no instrument so setting detector IDs makes no sense, but if
-      // we don't we break quite a few unit tests relying on this legacy
-      // behavior.
-      outputWS->getSpectrum(local_i).setDetectorID(i + 1);
```
Comment (OwenArnold) [inline Framework/Algorithms/src/CreateWorkspace.cpp]: Good to see this gone. Could you modify the unit tests for `CreateWorkspace` to guard against its accidental/deliberate reintroduction

File: Framework/API/src/ExperimentInfo.cpp [hunk 3]
```diff
@@ -491,16 +491,12 @@ void ExperimentInfo::setNumberOfDetectorGroups(const size_t count) const {
 void ExperimentInfo::setDetectorGrouping(
```
Comment (OwenArnold) [inline Framework/API/src/ExperimentInfo.cpp]: This method is public, but as far as I can tell it is not used externally. It's called by `cacheDetectorGroupings`, which is only used by `ConvertToMD`, basically I'm wondering if the try-catch is needed anymore.

**Review by OwenArnold — COMMENTED** [review]
My major concern here is what we do now that the behaviour of `CreateWorkspace` has changed. You have correctly fixed things in our code and executed scripts. What about external scripts? How do we announce these changes to the community.

File: Framework/Algorithms/src/AsymmetryCalc.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply Framework/Algorithms/src/AsymmetryCalc.cpp]: Well, I think that is exactly the problem: `setSpectrumDefinition` will not let you set anything that corresponds to detector IDs that do not exist.

File: Framework/API/src/ExperimentInfo.cpp (hunk 3 above)
Comment (SimonHeybrock) [reply Framework/API/src/ExperimentInfo.cpp]: It strips out any bad detector IDs and we still have many places that set bad IDs so this cannot be removed.

**Comment by SimonHeybrock** [comment]
@OwenArnold Regarding your main concern: I agree that we should definitely announce this. Personally I see this as a bug fix since it removes potentially unwanted mappings. Anyone who actually needs sensible mapping to the instrument is like to run, e.g., `LoadInstrument` with `RewriteSpectrumMapping=True`, since the default IDs from `CreateWorkspace` are not sensible (or just match accidentally).

Would having this in the release notes be enough?

File: Framework/API/src/ExperimentInfo.cpp (hunk 3 above)
Comment (OwenArnold) [reply Framework/API/src/ExperimentInfo.cpp]: Ok. I think that is probably true. So lets leave this alone.

**Review by OwenArnold — APPROVED** [review]
Latest changes to documentation look good. Along the lines of our discussion this morning and sufficiently detailed too.

**Pull request merged**